CRADLE/correctbiasutils/__init__.py

```diff
@@ -275,101 +286,68 @@ def getResultBWHeader(regions, ctrlBWName):
-	# 	],
-	#	[
-	# 		['experiFile0Temp0.msl', 'experiFile0Temp1.msl', 'experiFile0Temp2.msl'],
-	# 		['experiFile1Temp0.msl', 'experiFile1Temp1.msl', 'experiFile1Temp2.msl']
-	# 	],
-	# ]
-	#
-	# Which matches what we want to do with the files better -- combine all temp files for a particular file together
-	# into a single BigWig file
-
-	ctrlFiles = [[] for _ in range(len(ctrlBWNames))]
-	experiFiles = [[] for _ in range(len(experiBWNames))]
-	for jobFiles in tempFiles:
-		jobCtrl = jobFiles[0]
-		jobExperi = jobFiles[1]
-		for i, ctrlFile in enumerate(jobCtrl):
-			ctrlFiles[i].append(ctrlFile)
-		for i, experiFile in enumerate(jobExperi):
-			experiFiles[i].append(experiFile)
-
-	return ctrlFiles, experiFiles
-
-def mergeBWFiles(outputDir, header, tempFiles, ctrlBWNames, experiBWNames):
-	ctrlFiles, experiFiles = rotateBWFileArrays(tempFiles, ctrlBWNames, experiBWNames)
 
+def outputBWFile(outputDir, filename):
+	signalBWName = '.'.join(filename.rsplit('/', 1)[-1].split(".")[:-1])
+	return os.path.join(outputDir, signalBWName + "_corrected.bw")
+
+
+def outputCorrectedTmpFile(outputDir, chromo, chromoId, filename):
+	signalBWName = '.'.join(filename.rsplit('/', 1)[-1].split(".")[:-1])
+	return os.path.join(outputDir, f"{signalBWName}_{chromo}_{chromoId}_corrected.tmp")
+
+
+def outputNormalizedTmpFile(outputDir, filename):
+	normObBWName = '.'.join(filename.rsplit('/', 1)[-1].split(".")[:-1])
+	return os.path.join(outputDir, normObBWName + "_normalized.tmp")
+
+
+def mergeBWFiles(outputDir, header, fileChromoInfo, ctrlBWNames, experiBWNames):
 	jobList = []
-	for i, ctrlFile in enumerate(ctrlBWNames):
-		jobList.append((ctrlFiles[i], header, outputBWFile(outputDir, ctrlFile)))
-	for i, experiFile in enumerate(experiBWNames):
-		jobList.append((experiFiles[i], header, outputBWFile(outputDir, experiFile)))
+	for ctrlFile in ctrlBWNames:
+		jobList.append((ctrlFile, header, fileChromoInfo, outputBWFile(outputDir, ctrlFile), outputDir))
+	for experiFile in experiBWNames:
+		jobList.append((experiFile, header, fileChromoInfo, outputBWFile(outputDir, experiFile), outputDir))
 
 	return process(len(ctrlBWNames) + len(experiBWNames), mergeCorrectedFilesToBW, jobList)
 
-def outputBWFile(outputDir, filename):
-	signalBWName = '.'.join(filename.rsplit('/', 1)[-1].split(".")[:-1])
-	return os.path.join(outputDir, signalBWName + "_corrected.bw")
 
-def mergeCorrectedFilesToBW(tempFiles, bwHeader, signalBWName):
+def mergeCorrectedFilesToBW(replicateFile, bwHeader, fileChromoInfo, signalBWName, outputDir):
 	signalBW = pyBigWig.open(signalBWName, "w")
 	signalBW.addHeader(bwHeader)
-	dataReadSize = struct.calcsize(CORRECTED_RC_TEMP_FILE_STRUCT_FORMAT) * 1_000_000 # a somewhat arbitrary choice
-
-	for tempFile in tempFiles:
-		with open(tempFile, 'rb') as dataFile:
+	dataReadSize = struct.calcsize(CORRECTED_RC_TEMP_FILE_STRUCT_FORMAT) * MERGE_FILES_BUFFER_SIZE
+	starts = np.zeros(MERGE_FILES_BUFFER_SIZE, dtype=np.int32)
```

Comment:
i wonder the reason for initializing the large arrays (`starts`, `ends`, `values`) as opposed to using an empty array and then adding values in the previous version

Reply:
Creating an array of the max size once and inserting into it will be faster than creating many new arrays and appending to them.

Reply:
plus numpy arrays are smaller than python lists.